presto-main/src/main/java/com/facebook/presto/sql/gen/ExpressionProfiler.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.gen;
+
+import com.facebook.presto.annotation.UsedByGeneratedCode;
+import com.google.common.annotations.VisibleForTesting;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import static com.google.common.base.Verify.verify;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+/**
+ * This class can be called from generated projection classes to profile a single
+ * expression over many rows, and determines whether the profiled expression is expensive.
+ * That information is used by the projection class to determine whether it should check
+ * the yield signal in the projection tight loop. Please see {@link PageFunctionCompiler} for how to
+ * use this class in generated code.
+ */
+@UsedByGeneratedCode
+@NotThreadSafe
+public class ExpressionProfiler
+{
+    private static final int NUMBER_OF_ROWS_TO_PROFILE = 1024;
+    private static final int EXPENSIVE_FUNCTION_THRESHOLD_MILLIS = 1_000;
```

Comment:
if the function takes below this threshold (eg 0.5s), do we invoke it for all positions within a page without yielding? (eg. 1000 times)
what would be the max time before yielding?

Reply:
Yes, thread will return after `evaluate` is called for all positions in a page. That's why I mentioned above in `Other Notes` that I am not sure whether this threshold is a good one or not. In the worst case if each `evaluate` takes slightly below 1s (so it's a cheap function) for a page of 1024 positions it will take 1*1024/60 = 17m before the thread returns.